File: 백준/Gold/17136. 색종이 붙이기/색종이 붙이기.py
board = [list(map(int,input().split())) for _ in range(10)]
#색종이 사용 갯수
paper = [0,0,0,0,0]

def dfs(x,y,cnt):
    global min_val
    
    #y좌표 넘어가면 끝
    if y >= 10:
        min_val = min(min_val,cnt)
        return
    
    #x좌표 넘어가면 다음 y좌표
    if x >= 10:
        dfs(0,y+1,cnt)
        return
    
    if board[x][y] == 1:
        #5종류 색종이 시행
        for k in range(5):
            #최대 5장까지 가능
            if paper[k] == 5:
                continue
            if x+k >= 10 or y+k >= 10:
                continue
            
            #0체크
            flag = False
            for i in range(x,x+k+1):
                for j in range(y,y+k+1):
                    #0이 있으면 멈춤
                    if board[i][j] == 0:
                        flag = True
                        break
                if flag:
                    break
            
            #다 1이면 색종이 한장 추가하고 0으로 바꿔주기
            if flag == False:
                for i in range(x,x+k+1):
                    for j in range(y,y+k+1):
                        board[i][j] = 0
                        
                paper[k] +=1
                dfs(x+k+1,y,cnt+1)
                paper[k] -=1
                
                for i in range(x,x+k+1):
                    for j in range(y,y+k+1):
                        board[i][j] = 1
    
    else:
        dfs(x+1,y,cnt)
                

min_val = 1e9
dfs(0,0,0)
if min_val == 1e9:
    print(-1)
else:
    print(min_val)

File: 백준/Gold/17136. 색종이 붙이기/색종이 붙이기.java
import java.util.Scanner;

public class Main {
    static int[][] board = new int[10][10];
    static int[] paper = new int[5];

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Inputting board
        for (int i = 0; i < 10; i++) {
            for (int j = 0; j < 10; j++) {
                board[i][j] = scanner.nextInt();
            }
        }

        int result = dfs(0, 0, 0);
        if (result == 1e9) {
            System.out.println(-1);
        } else {
            System.out.println(result);
        }
    }

    static int dfs(int x, int y, int cnt) {
        int minVal = 1_000_000_000;

        if (y > 9) {
            minVal = Math.min(minVal, cnt);
            return minVal;
        }

        if (x > 9) {
            return dfs(0, y + 1, cnt);
        }

        if (board[x][y] == 1) {
            for (int k = 0; k < 5; k++) {
                if (paper[k] == 5) {
                    continue;
                }
                if (x + k > 9 || y + k > 9) {
                    continue;
                }

                boolean flag = false;
                for (int i = x; i <= x + k; i++) {
                    for (int j = y; j <= y + k; j++) {
                        if (board[i][j] == 0) {
                            flag = true;
                            break;
                        }
                    }
                    if (flag) {
                        break;
                    }
                }

                if (!flag) {
                    for (int i = x; i <= x + k; i++) {
                        for (int j = y; j <= y + k; j++) {
                            board[i][j] = 0;
                        }
                    }
                    paper[k]++;
                    minVal = Math.min(minVal, dfs(x + k + 1, y, cnt + 1));
                    paper[k]--;

                    for (int i = x; i <= x + k; i++) {
                        for (int j = y; j <= y + k; j++) {
                            board[i][j] = 1;
                        }
                    }
                }
            }
        } else {
            minVal = dfs(x + 1, y, cnt);
        }
        return minVal;
    }
}